sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -128,14 +150,111 @@ static void *_event_loop(void *arg)
     }
 
     event_queue_init(&_queue);
-    sock_udp_event_init(&_sock_udp, &_queue, _on_sock_evt, NULL);
-    event_loop(&_queue);
 
+#ifdef CONFIG_GCOAP_ENABLE_DTLS
+    sock_dtls_init();
+    if (sock_dtls_create(&_sock_dtls, &_sock_udp,
+                        SOCK_DTLS_CLIENT_TAG,
+                        SOCK_DTLS_1_2, SOCK_DTLS_SERVER) < 0) {
+        puts("Error creating DTLS sock");
+        sock_udp_close(&_sock_udp);
+        return 0;
+    }
+
+    dtls_session_management_init();
+    sock_dtls_event_init(&_sock_dtls, &_queue, _on_sock_dtls_evt,
+                        NULL);
+#else
+    sock_udp_event_init(&_sock_udp, &_queue, _on_sock_udp_evt, NULL);
+#endif
+
+    event_loop(&_queue);
     return 0;
 }
 
+#ifdef CONFIG_GCOAP_ENABLE_DTLS
+
+/* Handles DTLS socket events from the event queue */
+static void _on_sock_dtls_evt(sock_dtls_t *sock, sock_async_flags_t type, void *arg) {
+    (void)arg;
+    sock_dtls_session_t session = { 0 };
+
+    if (type & SOCK_ASYNC_CONN_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &session,
+                            _listen_buf, sizeof(_listen_buf),
+                            CONFIG_GCOAP_DTLS_HANDSHAKE_TIMEOUT_USEC);
+        if (res != -SOCK_DTLS_HANDSHAKE) {
+            DEBUG("gcoap: could not establish DTLS session: %d\n", res);
+            sock_dtls_session_destroy(sock, &session);
+            return;
+        }
+
+        /* If session has been previously stored and the state was HANDSHAKE_ONGOING
+        before, the handshake has been initiated internally and another thread
+        is waiting for the handshake.   */
+        session_state_t prev_state = NO_SESSION;
+        res = dtls_session_store(&session, SESSION_ESTABLISHED, &prev_state, false);
+        if (res) {
+            /* Send message to main thread to inform about established session */
+            if (prev_state == HANDSHAKE_ONGOING) {
+                msg_t msg = { .type = DTLS_EVENT_CONNECTED };
+                msg_send(&msg, _main_thread);
+            }
+        } else {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &session);
+        }
+
+        /* if no session slot is left: set timeout to free up session */
```

Comment:
What's the reason for not calling `_dtls_free_up_session()` directly?

Reply:
If more than one session is allowed yes: than it would be okay. But if only one session is allowed you don't want to destroy the session immediately after it was established. some time for coap messaging is needed.